Allow configuring programs directory

ruuda opened the pull request:
This removes the hard-coded `$LOCALAPPDATA\Programs\stack` directory, and uses `$STACK_ROOT\programs` instead, unifying this path for all programs. This resolves #1644 and also [my comment](https://github.com/commercialhaskell/stack/issues/996#issuecomment-258642367) on #996.

@borsboom: You noted in #1644 that it would be good to make this configurable in `stack.yaml`. That would certainly be possible, but currently I do not see how that would be useful. I think it would introduce more knobs than necessary. Is there a compelling use case for downloading and extracting GHC to a different location than the snapshots?

* [x] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [x] The documentation has been updated, if necessary.

mgsloan commented:
Thanks for working on this!  The problem with this approach is that it means people will need to re-install their GHC versions, and will also leave the old GHC versions still there.  This is why we'd prefer to make this an explicit option.

The reason to allow it to be specified in the config subset of `stack.yaml` is that then users can specify it in their `config.yaml`

ruuda commented:
I see. In that case, I think the proper thing to do (from a user perspective) would be to automatically move the data to the new location. But that would mean adding even more special cases for a while. Making this a configurable option with different defaults on Windows than on other platforms looks like an easier solution from an implementation point of view. What do you think?

mgsloan commented:
> Making this a configurable option with different defaults on Windows than on other platforms looks like an easier solution from an implementation point of view.

Sounds good to me!

ruuda commented:
I rebased this and changed it to keep the `$LOCALAPPDATA` default. There now is a `local-programs-path`, which defaults to `$STACK_ROOT/programs` except on Windows, where it defaults to `$LOCALAPPDATA\Programs\stack`. When getting `$LOCALAPPDATA` fails (I don’t think that should ever happen), it now defaults to `$STACK_ROOT\programs` on Windows too.

I have not tested this on Windows yet, I’ll try that tomorrow. I did test this on Linux.

mgsloan commented:
LGTM, thanks!

wowofbob commented:
> I rebased this and changed it to keep the `$LOCALAPPDATA` default. There now is a `local-programs-path`, which defaults to `$STACK_ROOT/programs` except on Windows, where it defaults to `$LOCALAPPDATA\Programs\stack`. When getting `$LOCALAPPDATA` fails (I don’t think that should ever happen), it now defaults to `$STACK_ROOT\programs` on Windows too.
> 
> I have not tested this on Windows yet, I’ll try that tomorrow. I did test this on Linux.

This pretty much seems to work on windows!
